errors.go

```diff
@@ -0,0 +1,204 @@
+package errkit
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/luci/go-render/render"
+
+	"github.com/kanisterio/errkit/internal/caller"
+)
+
+var _ error = (*PureError)(nil)
+var _ error = (*Error)(nil)
+var _ json.Marshaler = (*Error)(nil)
+var _ interface {
+	Unwrap() error
+} = (*Error)(nil)
+
+// Make an aliases for errors.Is, errors.As, errors.Unwrap
+// To avoid additional imports
+var (
+	Is     = errors.Is
+	As     = errors.As
+	Unwrap = errors.Unwrap
+)
+
+type PureError struct {
+	error
+}
+
+func NewPureError(message string) *PureError {
+	return &PureError{
+		error: errors.New(message),
+	}
+}
+
+type Error struct {
+	error
+	function   string
+	file       string
+	lineNumber int
+	details    ErrorDetails
+	cause      error
+	baseTag    *int // detect shallow copies, not needed in json
+	clonedFrom *Error
+}
+
+// New returns an error with the given message.
+func New(message string, details ...any) *Error {
+	return newError(message, 3, details...)
+}
+
+// Wrap returns a new Error that has the given message and err as the cause.
+func Wrap(err error, message string, details ...any) *Error {
+	return wrap(err, message, details...)
+}
+
+// WithStack wraps the given error with a struct that when serialized to JSON will return
+// a JSON object with the error message and error stack data.
+//
+// Note, this function should only be used with error values. E.g.:
+//
+//	var ErrWellKnownError = errors.New("Well-known error")
+//	...
+//	if someCondition {
+//	    return kerrors.WithStack(ErrWellKnownError)
+//	}
+func WithStack(err error) *Error {
+	var message string
+	if kerr, ok := err.(*Error); ok {
+		// We shouldn't pass *kerrors.Error to this function, but this will
+		// protect us from the situation when someone used kerrors.New()
+		// instead of errors.New(). Otherwise, the error will be double encoded
+		// JSON.
+		message = kerr.Message()
+	} else {
+		message = err.Error()
+	}
+
+	e := newError(message, 3)
+	e.cause = err
+	return e
+}
+
+func wrap(err error, message string, details ...any) *Error {
+	e := newError(message, 4, details...)
+	e.cause = err
+	return e
+}
+
+func newError(message string, stackDepth int, details ...any) *Error {
+	c := caller.GetFrame(stackDepth)
+	e := &Error{
+		error:      errors.New(message),
+		function:   c.Function,
+		lineNumber: c.Line,
+		// line number is intentionally appended to the file name
+		// this reduces the time needed to read the info and
+		// simplifies the navigation in the IDEs.
+		file:    fmt.Sprintf("%s:%d", c.File, c.Line),
+		details: ToErrorDetails(details),
+	}
+	// shallow copies (e.g. WithField(s)) share baseTag pointer to otherwise unused int
+	tag := 42
+	e.baseTag = &tag
+	return e
+}
+
+// WithCause adds a cause to the given error.
```

Comment:
Could this be better documented? It took me some time to figure out what's going on here.
It does have quite complex and confusing behaviour. What's the use of it?

Reply:
Explained (including an example) in readme. Do you think we need to put the same info here ?

Reply:
I see that, but I still have some issues with how this is different from `Wrap`:
- Why does it return `error` and not `*Error`?
- It's using the error passed as the `error` field, which is not what `Wrap` is doing, also this field is only used in `Is` function which is inconsistent with `errors.Is`, as mentioned in other comment.

Reply:
Returning a concrete type as a signature may have an undesired effect, when a nil error won't be nil, because nil pointer to a concrete type is not equal to nil in go. If I was not clear enough, I can share an example.

Reply:
Thanks for pointing on inconsistency in interfaces. Now interfaces are consistent and all of them are returning `error`.
`Is` function on `errkitError` is needed to make `errors.Is` work properly and not intended to be used directly.

Reply:
> Is function on errkitError is needed to make errors.Is work properly and not intended to be used directly.
Ok. That makes sense. Maybe we can put some warning next to the function definition to highlight that.

Current implementation of `WithStack` and `WithCause` raises some questions:
- Why do we create a new error in `WithStack` instead of just putting existing error in the `error` field? We do that in `WithCause` already.
- For both functions there is a note that they should not be used with errkit errors, but since we don't force it in compile time there is a workaround which breaks the contract (such as `Is` matching). Also as a user of the code I might actually use them intentionally or unintentionally. Maybe we can come up with some reasonable implementations for those?

Reply:
Thanks for pointing, there is really no reason to create new error in `WithStack`. 
The second point is more complex, let's discuss it on our today call.